### ISSUE TITLE
How to reuse the thread pool in different places in the code?

### ISSUE THREAD

**Issue by Sven-v-Beuningen** [taskflow/taskflow#382]
Hi,

I'm currently integrating TaskFlow into an existing rendering engine. I came across a situation, where I think I'm doing it wrong and could need an advise.

Starting from my main loop I have a complex TaskFlow which triggers all subsystems of the rendering engine. The execution graph takes care of all dependencies. This works perfectly fine so far.

One of the subsystems is updating the scene which is rendered. Traversing the scene and doing calculations can be easily done in parallel. Therefore I used tf::Subflow to achieve this. This also does work like expected.
Because the traversing of the scene is located deeper in the execution stack I need to forward the Subflow from the main loop, several steps down to the traversing of the scene. This doesn't feel very clean and anoying. It leads to pushing subflow forward through some classes which do not need the Subflow at all. 

To clean this up, I tried to not push the subflow forward, but create another tf::TaskFlow right at the place where I do the traversing of the scene. This TaskFlow is then executed with the same Executor to reuse the threadpool. Unfortunately this seems to lead to a deadlock. I assume the reason is, that I call Executor::run from within an already running task from the main loop. 

Could this be the case? If yes, what is the right way to reuse the thread pool from different places in the code without pushing subflows through the stack? Fun fact, this only does happen in release build. The debug build does work fine.

Best Regards

Sven

**Comment by Sven-v-Beuningen** [comment]
I was able make a minimal example on this:

```
    tf::Executor      executor;
    tf::Taskflow      tf1;

    tf1.emplace([&](tf::Subflow& subflow) { 
        // some imaginary levels down in call stack

        // loop counter has to be at least number of logical cpus
        for (auto i = 0; i < 12; ++i)
        {
            subflow.emplace([&]() {
                tf::Taskflow tf2;

                tf2.emplace([&](tf::Subflow& sf) { sf.emplace([&]() { }); });
                executor.run(tf2).wait();
            });
        }
    });

    executor.run(tf1).wait();
```

**Comment by Sven-v-Beuningen** [comment]
Does anybody have an Idea how to solve this? Or am I the only one who uses TaskFlow in different places in a huge project?

**Comment by tsung-wei-huang** [comment]
@Sven-v-Beuningen Sorry for the late response! I thought my message in the draft was sent - in your case, you should try avoiding calling `"wait"` from a worker in the executor. This will cause that worker to block waiting until the job finishes, in your case, `executor.run(tf2).wait()`. If you want to do this, subflow is the way to go.

**Comment by Sven-v-Beuningen** [comment]
@tsung-wei-huang Thank you for the reply. I need to wait, because I have to make sure all calculations are finished, before I move on to the next task. So the only way to achieve what I need seems to push forward the subflow over and over again down the call stack.

**Comment by tsung-wei-huang** [comment]
Yes. That is the only way regarding your workload so far. Another solution is to propose fiber (co-routine)-based future and integrate them into the executor, but this requires further future work. 

Will close the issue for now. Feel free to reopen it or post another issue if you have other questions.
